apps/definition_kafka/lib/kafka/topic/destination.ex

```diff
@@ -0,0 +1,174 @@
+      Ok.error(reason)
+    end
+  end
+
+  @impl GenServer
+  def init({topic, context}) do
+    Process.flag(:trap_exit, true)
+    state = Map.from_struct(context) |> Map.put(:connection, connection_name())
+    {:ok, state, {:continue, {:init, topic}}}
+  end
+
+  @impl GenServer
+  def handle_continue({:init, topic}, state) do
+    with :ok <- Elsa.create_topic(topic.endpoints, topic.name, partitions: topic.partitions),
+         {:ok, pid} <- start_producer(topic, state.connection) do
+      store_connection(topic, state.connection)
+      {:noreply, Map.put(state, :elsa_pid, pid)}
+    end
+  end
+
+  @impl GenServer
+  def handle_call(:stop, _from, state) do
+    Logger.info(fn -> "#{__MODULE__}: Terminating by request" end)
+    {:stop, :normal, :ok, state}
+  end
+
+  @impl GenServer
+  def handle_cast({:dlq, []}, state) do
+    {:noreply, state}
+  end
+
+  def handle_cast({:dlq, messages}, state) do
+    Logger.debug(fn -> "#{__MODULE__}: Writing #{Enum.count(messages)} messages to DLQ" end)
+
+    opts = Enum.into(state, [])
+
+    dead_letters =
+      Enum.map(messages, fn {og, reason} ->
+        Keyword.merge(opts, reason: reason, original_message: og)
+        |> DeadLetter.new()
+      end)
+
+    dlq().write(dead_letters)
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info({:EXIT, pid, reason}, %{elsa_pid: pid} = state) do
+    Logger.error(fn -> "#{__MODULE__}: Elsa(#{inspect(pid)}) died : #{inspect(reason)}" end)
+    {:stop, reason, state}
+  end
+
+  def handle_info(message, state) do
+    Logger.info(fn -> "#{__MODULE__}: received unknown message - #{inspect(message)}" end)
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def terminate(reason, %{elsa_pid: pid}) do
+    Process.exit(pid, reason)
+
+    receive do
+      {:EXIT, ^pid, _} -> reason
+    after
+      20_000 -> reason
+    end
+  end
+
+  def terminate(reason, _) do
+    reason
+  end
+
+  defp start_producer(topic, conn) do
+    with {:ok, pid} <-
+           Elsa.Supervisor.start_link(
+             connection: conn,
+             endpoints: topic.endpoints,
+             producer: [
+               topic: topic.name
+             ]
+           ),
+         true <- Elsa.Producer.ready?(conn) do
+      Ok.ok(pid)
+    end
+  end
+
+  defp do_write(topic, messages) do
+    connection(topic)
+    |> Ok.map(&Elsa.produce(&1, topic.name, messages, partitioner: topic.partitioner))
+  end
+
+  @retry with: constant_backoff(100) |> take(10)
+  defp connection(topic) do
+    table_name(topic)
+    |> :ets.lookup_element(topic.pid, 2)
+    |> Ok.ok()
+  catch
+    _, reason ->
+      Ok.error(reason)
+  end
+
+  defp store_connection(topic, connection) do
+    table = table_name(topic)
+
+    case :ets.whereis(table) do
+      :undefined -> :ets.new(table, [:named_table, :protected])
+      _ -> :ok
+    end
+
+    :ets.insert(table, {self(), connection})
+  end
```

Comment:
since this is a protected table, no one else can write to it but the process that was just init'ed; since this is in a handle_continue where would this table have come from if it were to already exist?

Reply:
That... is a fair point. I copied this logic from some testing modules without really thinking it through. There, it makes sense. This does not.